From f01f6a847474755ee00a86c599cda3cbadef34fe Mon Sep 17 00:00:00 2001
From: Justin Ruggles <justin.ruggles@gmail.com>
Date: Mon, 9 Jan 2012 13:49:52 -0500
Subject: [PATCH 06/63] gsm demuxer: use generic seeking instead of a
 gsm-specific function.

Since packets all contain only a single block, the generic seek function can
be used while still maintaining block-accuracy.
---
 libavformat/gsmdec.c |   35 +----------------------------------
 1 files changed, 1 insertions(+), 34 deletions(-)

diff --git a/libavformat/gsmdec.c b/libavformat/gsmdec.c
index a23ce3b..443f820 100644
--- a/libavformat/gsmdec.c
+++ b/libavformat/gsmdec.c
@@ -72,39 +72,6 @@ static int gsm_read_header(AVFormatContext *s, AVFormatParameters *ap)
     return 0;
 }
 
-static int gsm_read_seek2(AVFormatContext *s, int stream_index, int64_t min_ts,
-                          int64_t ts, int64_t max_ts, int flags)
-{
-    GSMDemuxerContext *c = s->priv_data;
-
-    /* convert timestamps to file positions */
-    if (!(flags & AVSEEK_FLAG_BYTE)) {
-        if (stream_index < 0) {
-            AVRational bitrate_q = { GSM_BLOCK_SAMPLES, c->sample_rate * GSM_BLOCK_SIZE };
-            ts     = av_rescale_q(ts,     AV_TIME_BASE_Q, bitrate_q);
-            min_ts = av_rescale_q(min_ts, AV_TIME_BASE_Q, bitrate_q);
-            max_ts = av_rescale_q(max_ts, AV_TIME_BASE_Q, bitrate_q);
-        } else {
-            ts     *= GSM_BLOCK_SIZE;
-            min_ts *= GSM_BLOCK_SIZE;
-            max_ts *= GSM_BLOCK_SIZE;
-        }
-    }
-    /* round to nearest block boundary */
-    ts = (ts + GSM_BLOCK_SIZE / 2) / GSM_BLOCK_SIZE * GSM_BLOCK_SIZE;
-    ts = FFMAX(0, ts);
-
-    /* handle min/max */
-    while (ts < min_ts)
-        ts += GSM_BLOCK_SIZE;
-    while (ts > max_ts)
-        ts -= GSM_BLOCK_SIZE;
-    if (ts < min_ts || ts > max_ts)
-        return -1;
-
-    return avio_seek(s->pb, ts, SEEK_SET);
-}
-
 static const AVOption options[] = {
     { "sample_rate", "", offsetof(GSMDemuxerContext, sample_rate),
        AV_OPT_TYPE_INT, {.dbl = GSM_SAMPLE_RATE}, 1, INT_MAX / GSM_BLOCK_SIZE,
@@ -125,7 +92,7 @@ AVInputFormat ff_gsm_demuxer = {
     .priv_data_size = sizeof(GSMDemuxerContext),
     .read_header    = gsm_read_header,
     .read_packet    = gsm_read_packet,
-    .read_seek2     = gsm_read_seek2,
+    .flags          = AVFMT_GENERIC_INDEX,
     .extensions     = "gsm",
     .value          = CODEC_ID_GSM,
     .priv_class     = &class,
-- 
1.7.5.4

